### PR TITLE
Change Entity::generation from u32 to NonZeroU32 for niche optimization

### PR DESCRIPTION
**Pull request by notverymoe**
# Objective

- Implements change described in https://github.com/bevyengine/bevy/issues/3022
- Goal is to allow Entity to benefit from niche optimization, especially in the case of Option<Entity> to reduce memory overhead with structures with empty slots

## Discussion
- First PR attempt: https://github.com/bevyengine/bevy/pull/3029
- Discord: https://discord.com/channels/691052431525675048/1154573759752183808/1154573764240093224

## Solution

- Change `Entity::generation` from u32 to NonZeroU32 to allow for niche optimization.
  - The reason for changing generation rather than index is so that the costs are only encountered on Entity free, instead of on Entity alloc
  - There was some concern with generations being used, due to there being some desire to introduce flags. This was more to do with the original retirement approach, however, in reality even if generations were reduced to 24-bits, we would still have 16 million generations available before wrapping and current ideas indicate that we would be using closer to 4-bits for flags.
  - Additionally, another concern was the representation of relationships where NonZeroU32 prevents us using the full address space, talking with Joy it seems unlikely to be an issue. The majority of the time these entity references will be low-index entries (ie. `ChildOf`, `Owes`), these will be able to be fast lookups, and the remainder of the range can use slower lookups to map to the address space.  
  - It has the additional benefit of being less visible to most users, since generation is only ever really set through `from_bits` type methods.
- `EntityMeta` was changed to match
- On free, generation now explicitly wraps:
  - Originally, generation would panic in debug mode and wrap in release mode due to using regular ops.
  - The first attempt at this PR changed the behavior to "retire" slots and remove them from use when generations overflowed. This change was controversial, and likely needs a proper RFC/discussion.
  - Wrapping matches current release behaviour, and should therefore be less controversial.
  - Wrapping also more easily migrates to the retirement approach, as users likely to exhaust the exorbitant supply of generations will code defensively against aliasing and that defensive code is less likely to break than code assuming that generations don't wrap.
  - We use some unsafe code here when wrapping generations, to avoid branch on NonZeroU32 construction. It's guaranteed safe due to how we perform wrapping and it results in significantly smaller ASM code.
    - https://godbolt.org/z/6b6hj8PrM 

## Migration

- Previous `bevy_scene` serializations have a high likelihood of being broken, as they contain 0th generation entities.

## Current Issues
 
- `Entities::reserve_generations` and `EntityMapper` wrap now, even in debug - although they technically did in release mode already so this probably isn't a huge issue. It just depends if we need to change anything here?

### CODE REVIEW

**Comment by A-Walrus** [comment]
We might want to write a little migration script that increases the generation of every entity in a scene file by one

**Review by rodolphito — APPROVED** [review]
appears sensible, should probably merge after #9797 though

**Comment by notverymoe** [comment]
> We might want to write a little migration script that increases the generation of every entity in a scene file by one

Do we have any prior art for this? Be good to look at before I give it a go

**Comment by notverymoe** [comment]
> appears sensible, should probably merge after #9797 though

Seems like this'll need a rework after that lands, which is fine - I'll keep my eyes out for it. I'll look at doing the benchmarking after that's in and I fix things up.

**Comment by Kolsky** [comment]
> The wrapping stuff

If you want to do the usual wrapping for addition of two integers, here is how ([playground](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=a8eb8b310e523a6feca627b05788b125))

Proof that it works: adding two integers will result in exactly 0 or 1 overflows (2 × uN::MAX = (oveflow_flag, uN::MAX - 1)). The case with 0 overflows is trivial (just add them normally). If overflow happens, the result would be off-by-one with the normal wrapping. There are two extreme cases: if result was uN::MIN, adding 1 will make it NonZeroUN::MIN. If result was uN::MAX - 1, adding 1 will make it NonZeroUN::MAX. Thus, neither underflow nor overflow of NonZero domain is possible. There's also a commented out simple test for Miri, albeit it needs to be run locally.

Tl;dr: it's as simple as (x + y + overflow_flag)

**Comment by notverymoe** [comment]
@Kolsky Oh, that's really excellent :O Thanks for that!

**Comment by notverymoe** [comment]
The optimized assembly is also really clean, nice!

**Comment by notverymoe** [comment]
Still waiting on that other PR to merge, but I ran bench with 857fb9c72:
[bench.txt](https://github.com/bevyengine/bevy/files/12774232/bench.txt)

**Comment by notverymoe** [comment]
I'm not too familiar with those numbers, but they seem rather inconsistent between different magnitudes of test:
X5950, 32GiB ram, Running EndeavourOS with Linux 6.5.4

Shut everything down on the machine except the terminal running the test.

**Comment by alice-i-cecile** [comment]
Yeah, the benchmarks are definitely a bit noisy, both here and in general 🤔 I can't make clear sense of this unfortunately.

**Comment by notverymoe** [comment]
Elabajaba on discord pointed out that since I'm on a Zen 3 CPU, I'm probably getting different boosts on every core. I'll try disabling that and running the tests again, see if I get something more stable and less noisey.

**Comment by notverymoe** [comment]
Yep, that was it. I disabled Precision Boost and locked my CPU scaling to 2.2GHz on linux. Much more stable, not as much swing between tests, still a little noisey. But I think maybe i need to run a few more tests, because the results are little strange in places (30% gain on events_iter??)

Ran against bb13d065d on main, and merged into the branch, threshold of 5%:
```
group                                           main                    niche_ecs
-----                                           ------                  -----------
added_archetypes/archetype_count/100            1.35     59.2±1.25µs    1.00     43.8±0.33µs
added_archetypes/archetype_count/1000           1.14   1059.4±5.62µs    1.00    930.6±3.45µs
added_archetypes/archetype_count/10000          1.18     12.8±0.53ms    1.00     10.9±0.27ms
added_archetypes/archetype_count/200            1.24    108.4±2.56µs    1.00     87.4±4.70µs
added_archetypes/archetype_count/2000           1.13      2.2±0.02ms    1.00   1957.9±9.80µs
added_archetypes/archetype_count/500            1.22   413.1±29.71µs    1.00   338.7±22.49µs
added_archetypes/archetype_count/5000           1.16      5.9±0.03ms    1.00      5.1±0.02ms
build_schedule/1000_schedule                    1.08       3.6±0.02s    1.00       3.3±0.04s
build_schedule/1000_schedule_noconstraints      1.20     32.0±0.27ms    1.00     26.5±0.35ms
build_schedule/100_schedule                     1.09     22.0±0.06ms    1.00     20.3±0.04ms
build_schedule/500_schedule                     1.09   687.7±10.20ms    1.00    633.1±7.57ms
busy_systems/01x_entities_06_systems            1.00     72.9±4.26µs    1.08     78.4±4.59µs
busy_systems/01x_entities_12_systems            1.00    131.8±3.45µs    1.05    138.9±3.91µs
busy_systems/01x_entities_15_systems            1.00    164.4±4.84µs    1.06    173.7±7.20µs
busy_systems/02x_entities_06_systems            1.00    120.8±2.33µs    1.06    127.5±3.24µs
busy_systems/05x_entities_03_systems            1.00    139.9±1.11µs    1.13    158.4±3.67µs
busy_systems/05x_entities_06_systems            1.00    256.1±1.77µs    1.16    296.3±1.77µs
busy_systems/05x_entities_09_systems            1.00    379.2±2.08µs    1.15    437.7±6.34µs
busy_systems/05x_entities_12_systems            1.00    502.6±6.63µs    1.15    575.8±7.26µs
busy_systems/05x_entities_15_systems            1.00    622.2±8.26µs    1.15    713.9±8.03µs
contrived/01x_entities_06_systems               1.12     58.7±4.38µs    1.00     52.4±2.42µs
contrived/01x_entities_15_systems               1.08    117.7±9.58µs    1.00    109.3±8.53µs
contrived/03x_entities_03_systems               1.00     54.0±3.66µs    1.11     60.2±4.75µs
contrived/03x_entities_09_systems               1.08   134.5±10.70µs    1.00    124.6±3.17µs
contrived/04x_entities_06_systems               1.06    113.1±5.34µs    1.00    106.6±3.01µs
contrived/05x_entities_09_systems               1.06   192.8±11.52µs    1.00    182.3±7.92µs
empty_systems/002_systems                       1.25     18.5±1.74µs    1.00     14.8±0.36µs
empty_systems/025_systems                       1.05     34.8±2.43µs    1.00     33.1±1.79µs
empty_systems/030_systems                       1.00     37.5±1.44µs    1.07     40.3±5.48µs
empty_systems/035_systems                       1.00     42.7±5.36µs    1.25    53.6±11.26µs
empty_systems/040_systems                       1.33    64.1±12.02µs    1.00     48.0±7.18µs
empty_systems/045_systems                       1.33    65.8±13.64µs    1.00     49.5±0.81µs
empty_systems/050_systems                       1.12    69.1±12.46µs    1.00    61.8±11.82µs
empty_systems/055_systems                       1.13    78.0±16.51µs    1.00    69.3±11.69µs
empty_systems/060_systems                       1.08    80.5±13.07µs    1.00     74.5±9.37µs
empty_systems/065_systems                       1.12    86.8±13.46µs    1.00     77.7±9.74µs
empty_systems/070_systems                       1.00    95.0±15.66µs    1.07   101.4±13.63µs
empty_systems/075_systems                       1.09    97.3±14.35µs    1.00    89.4±11.86µs
empty_systems/080_systems                       1.11   114.7±16.54µs    1.00   103.2±19.28µs
empty_systems/085_systems                       1.00     97.7±7.42µs    1.17   114.0±18.74µs
events_iter/size_16_events_100                  1.45    146.6±0.13ns    1.00    100.8±0.51ns
events_iter/size_16_events_1000                 1.50   1380.8±5.29ns    1.00    920.5±0.44ns
events_iter/size_16_events_10000                1.50     13.7±0.03µs    1.00      9.1±0.00µs
events_iter/size_16_events_50000                1.50     68.4±0.25µs    1.00     45.7±0.50µs
events_iter/size_4_events_100                   1.46    146.7±0.17ns    1.00    100.8±0.53ns
events_iter/size_4_events_1000                  1.50  1381.9±30.73ns    1.00    920.4±0.59ns
events_iter/size_4_events_10000                 1.50     13.7±0.05µs    1.00      9.1±0.01µs
events_iter/size_4_events_50000                 1.50     68.6±0.11µs    1.00     45.6±0.14µs
events_send/size_16_events_100                  1.00    238.7±0.42ns    1.14    271.7±0.57ns
events_send/size_512_events_100                 1.00      2.6±0.00µs    1.06      2.8±0.01µs
iter_fragmented/base                            1.00   799.9±24.48ns    1.29  1030.2±24.57ns
iter_fragmented_sparse/wide                     1.13     85.2±6.57ns    1.00    75.5±18.44ns
iter_simple/system                              1.64     22.9±0.03µs    1.00     14.0±0.07µs
iter_simple/wide                                1.00     65.1±0.88µs    1.13     73.8±4.25µs
query_get/50000_entities_table                  1.08    476.5±1.61µs    1.00    440.5±4.31µs
query_get_component/50000_entities_table        1.07   1039.1±7.86µs    1.00    966.6±8.49µs
query_get_many_2/50000_calls_table              1.07    835.6±7.38µs    1.00    780.8±1.99µs
run_condition/no/021_systems                    1.00     14.9±0.15µs    1.08     16.1±1.63µs
run_condition/no/026_systems                    1.00     15.7±1.47µs    1.10     17.3±1.16µs
run_condition/no/041_systems                    1.00     16.5±0.90µs    1.06     17.5±1.65µs
run_condition/no/066_systems                    1.00     17.2±1.37µs    1.14     19.7±0.58µs
run_condition/no/071_systems                    1.08     20.1±0.63µs    1.00     18.6±1.81µs
run_condition/no/076_systems                    1.00     17.1±0.64µs    1.23     21.1±2.06µs
run_condition/no/081_systems                    1.00     19.0±1.54µs    1.06     20.2±1.68µs
run_condition/no/086_systems                    1.00     20.0±1.60µs    1.06     21.3±0.41µs
run_condition/no/096_systems                    1.21     22.3±0.73µs    1.00     18.4±0.67µs
run_condition/yes/016_systems                   1.13     31.1±4.65µs    1.00     27.6±2.89µs
run_condition/yes/031_systems                   1.00     39.3±5.34µs    1.07     42.0±7.37µs
run_condition/yes/036_systems                   1.05     45.1±3.84µs    1.00     42.8±2.24µs
run_condition/yes/051_systems                   1.00     63.0±7.82µs    1.05    66.4±12.62µs
run_condition/yes/056_systems                   1.00    70.1±10.10µs    1.15    80.7±14.06µs
run_condition/yes/061_systems                   1.00     74.6±8.93µs    1.07    80.1±15.08µs
run_condition/yes/066_systems                   1.15   104.7±15.10µs    1.00    90.7±16.92µs
run_condition/yes/071_systems                   1.00    89.3±11.59µs    1.13   100.6±22.13µs
run_condition/yes/076_systems                   1.00     89.4±9.15µs    1.12   100.1±14.48µs
run_condition/yes/081_systems                   1.06    99.5±11.61µs    1.00     93.9±8.29µs
run_condition/yes/086_systems                   1.40   142.7±14.29µs    1.00   102.3±10.79µs
run_condition/yes/091_systems                   1.12   150.2±16.58µs    1.00   133.8±20.46µs
run_condition/yes/096_systems                   1.00   126.7±17.67µs    1.19   150.3±17.23µs
run_condition/yes_using_query/016_systems       1.07     28.6±2.33µs    1.00     26.8±0.42µs
run_condition/yes_using_query/031_systems       1.00     40.6±6.55µs    1.07     43.7±7.94µs
run_condition/yes_using_query/036_systems       1.11     47.2±7.30µs    1.00     42.7±0.89µs
run_condition/yes_using_query/046_systems       1.12    64.8±13.27µs    1.00     58.0±6.96µs
run_condition/yes_using_query/051_systems       1.00     64.4±7.90µs    1.07    69.0±12.70µs
run_condition/yes_using_query/061_systems       1.00    82.2±11.63µs    1.06    87.5±14.79µs
run_condition/yes_using_query/066_systems       1.08    88.3±13.49µs    1.00     81.7±9.15µs
run_condition/yes_using_query/071_systems       1.00    91.4±12.91µs    1.08    98.4±11.83µs
run_condition/yes_using_query/081_systems       1.00   105.3±19.28µs    1.07   112.2±14.51µs
run_condition/yes_using_query/086_systems       1.29   130.5±22.39µs    1.00    101.2±7.04µs
run_condition/yes_using_query/091_systems       1.15   126.8±18.31µs    1.00   110.2±12.02µs
run_condition/yes_using_query/101_systems       1.08   138.4±18.16µs    1.00   128.7±21.49µs
run_condition/yes_using_resource/036_systems    1.00     49.7±9.44µs    1.12    55.6±11.58µs
run_condition/yes_using_resource/046_systems    1.00     58.2±7.93µs    1.15    67.0±13.45µs
run_condition/yes_using_resource/051_systems    1.14    67.9±11.02µs    1.00     59.7±5.49µs
run_condition/yes_using_resource/056_systems    1.00    70.1±10.68µs    1.16    81.1±14.84µs
run_condition/yes_using_resource/061_systems    1.00     72.3±6.30µs    1.10    79.4±11.07µs
run_condition/yes_using_resource/066_systems    1.08    93.1±16.91µs    1.00    86.1±12.60µs
run_condition/yes_using_resource/076_systems    1.14    117.5±9.96µs    1.00   103.3±18.74µs
run_condition/yes_using_resource/081_systems    1.06   118.9±17.52µs    1.00   112.4±14.63µs
run_condition/yes_using_resource/086_systems    1.15   124.9±17.57µs    1.00   108.7±16.91µs
run_condition/yes_using_resource/096_systems    1.00   119.5±15.52µs    1.17   139.9±18.14µs
sized_commands_0_bytes/2000_commands            1.17      7.5±0.01µs    1.00      6.4±0.00µs
sized_commands_0_bytes/4000_commands            1.14     14.6±0.01µs    1.00     12.8±0.01µs
sized_commands_0_bytes/6000_commands            1.16     22.3±0.60µs    1.00     19.3±0.02µs
sized_commands_0_bytes/8000_commands            1.14     29.6±0.03µs    1.00     25.9±0.23µs
world_query_get/50000_entities_sparse_wide      1.00    389.1±0.51µs    1.07   415.6±40.24µs
world_query_iter/50000_entities_table           1.34     91.3±0.19µs    1.00     68.3±0.03µs
```

Seems like gains on `added_archetype`, `contrived`, `events_iter` and losses on `busy_systems` and `events_send`.

File: crates/bevy_ecs/src/entity/map_entities.rs
Comment (notverymoe) [reply crates/bevy_ecs/src/entity/map_entities.rs:152]: Excellent point, I can also make this change in the `lib.rs` tests in a few places.

**Comment by notverymoe** [comment]
(Sorry about that double-push, doing it between compiles at work and thought I already ran CI locally... oops)

**Comment by notverymoe** [comment]
At this point, would it be worth re-basing on #9797 ahead of time?

**Comment by notverymoe** [comment]
Thanks Alice, I'll get on that soon

File: crates/bevy_ecs/src/entity/mod.rs
Comment (notverymoe) [reply crates/bevy_ecs/src/entity/mod.rs:916]: @Kolsky up above came up with it: 
https://github.com/bevyengine/bevy/pull/9907#issuecomment-1732964111

I was really surprised at it's elegance, definitely going to be using it in one or two places in my personal projects now that i know about it. I think, unfortunately, given my reading of: 
https://github.com/bevyengine/bevy/pull/9797#issuecomment-1786990405

it means we won't be able to keep it (I think)

Comment (Bluefinger) [reply crates/bevy_ecs/src/entity/mod.rs:916]: Actually, I think it we will be able to do something similar, but we'd need to wrap on a however many bits we have available for the generation segment. So at the moment, that would be 31 bits. So you'd have to do something like:

```rust
pub const fn nonzero_wrapping_high_increment(value: NonZeroU32) -> NonZeroU32 {
     let next_value = value.get().wrapping_add(1);
     // Mask the overflow bit
     let overflowed = (next_value & 0x8000_0000) >> 31;

    // Remove the overflow bit from the next value, but then add to it
    unsafe { NonZeroU32::new_unchecked((next_value & 0x7FFF_FFFF) + overflowed) }
}
```

As long as we know we are only incrementing by one each time, then it should still output fairly terse asm: https://rust.godbolt.org/z/PnYTPfGb6 Basically the same principle as before, just applied to wrapping on 31 bits (or whatever amount of bits we need later)

Comment (notverymoe) [reply crates/bevy_ecs/src/entity/mod.rs:916]: Oh I see, that's interesting :thinking: We can definitely use that for the regular increment, it was just nice that the this version worked for both slot incrementing and `Entities::reserve_generations` (which requires an arbitrary increment, but honestly might not even be correct).

Comment (notverymoe) [reply crates/bevy_ecs/src/entity/mod.rs:916]: Tried a version with `checked_add`, it has a similar instruction count, but is probably more expensive.

**Review by AxiomaticSemantics — COMMENTED** [review]
Perhaps 
```rust
#[inline(always)]
```
here? Minor nit, I'm not sure if how hot this code path gets, ignore if non-hot.

File: crates/bevy_ecs/src/entity/mod.rs
Comment (AxiomaticSemantics) [inline crates/bevy_ecs/src/entity/mod.rs:275]: Perhaps 
```rust
#[inline(always)]
```
here? Minor nit, I'm not sure how hot this code path gets, ignore if non-hot.

**Review by maniwani — APPROVED** [review]
This looks good to me. I don't really have anything to add since we discussed most concerns in the linked Discord discussion. It's cool that we even avoided adding a branch to `free`.

**Comment by Testare** [comment]
Couldn't find information on this in the migration guide

**Comment by notverymoe** [comment]
I'm deeply sorry for the inconvenience I've caused with this. I'll look into some solutions.

**Pull request merged**
